src/libostree/ostree-repo.c

```diff
@@ -2614,225 +2614,199 @@ _ostree_repo_read_bare_fd (OstreeRepo           *self,
-                       const char         *checksum,
-                       GInputStream      **out_input,
-                       GFileInfo         **out_file_info,
-                       GVariant          **out_xattrs,
-                       GCancellable       *cancellable,
-                       GError            **error)
+static gboolean
+repo_load_file_archive (OstreeRepo *self,
+                        const char         *checksum,
+                        GInputStream      **out_input,
+                        GFileInfo         **out_file_info,
+                        GVariant          **out_xattrs,
+                        GCancellable       *cancellable,
+                        GError            **error)
 {
-  gboolean found = FALSE;
-  g_autoptr(GInputStream) ret_input = NULL;
-  g_autoptr(GFileInfo) ret_file_info = NULL;
-  g_autoptr(GVariant) ret_xattrs = NULL;
-
-  OstreeRepoMode repo_mode = ostree_repo_get_mode (self);
-
+  struct stat stbuf;
   char loose_path_buf[_OSTREE_LOOSE_PATH_MAX];
-  _ostree_loose_path (loose_path_buf, checksum, OSTREE_OBJECT_TYPE_FILE, repo_mode);
+  _ostree_loose_path (loose_path_buf, checksum, OSTREE_OBJECT_TYPE_FILE, self->mode);
 
-  if (repo_mode == OSTREE_REPO_MODE_ARCHIVE_Z2)
-    {
-      int fd = -1;
-      struct stat stbuf;
-      g_autoptr(GInputStream) tmp_stream = NULL;
+  glnx_fd_close int fd = -1;
+  if (!ot_openat_ignore_enoent (self->objects_dir_fd, loose_path_buf, &fd,
+                                error))
+    return FALSE;
 
-      if (!ot_openat_ignore_enoent (self->objects_dir_fd, loose_path_buf, &fd,
+  if (fd < 0 && self->commit_stagedir_fd != -1)
+    {
+      if (!ot_openat_ignore_enoent (self->commit_stagedir_fd, loose_path_buf, &fd,
                                     error))
         return FALSE;
+    }
 
-      if (fd < 0 && self->commit_stagedir_fd != -1)
+  if (fd != -1)
+    {
+      if (!glnx_fstat (fd, &stbuf, error))
+        return FALSE;
+
+      g_autoptr(GInputStream) tmp_stream = g_unix_input_stream_new (glnx_steal_fd (&fd), TRUE);
+      /* Note return here */
+      return ostree_content_stream_parse (TRUE, tmp_stream, stbuf.st_size, TRUE,
+                                          out_input, out_file_info, out_xattrs,
+                                          cancellable, error);
+    }
+  else
+    {
+      if (self->parent_repo)
         {
-          if (!ot_openat_ignore_enoent (self->commit_stagedir_fd, loose_path_buf, &fd,
-                                        error))
-            return FALSE;
+          return ostree_repo_load_file (self->parent_repo, checksum,
+                                        out_input, out_file_info, out_xattrs,
+                                        cancellable, error);
         }
-
-      if (fd != -1)
+      else
         {
-          tmp_stream = g_unix_input_stream_new (fd, TRUE);
-          fd = -1; /* Transfer ownership */
+          g_set_error (error, G_IO_ERROR, G_IO_ERROR_NOT_FOUND,
```

Comment:
Can't, we use `G_IO_ERROR_NOT_FOUND`.  (Though, maybe we could set `errno = ENOENT; return glnx_throw_errno()` but let's not scope that in now?)